File: converters/rust/src/directive_filter.rs
//! Directive filter framework for GraphQL SDL generation.
//!
//! Provides configuration-driven filtering of GraphQL directives
//! during SDL output, supporting modes like "viewer-friendly"
//! (strip all infrastructure directives) and "exclude-draft"
//! (omit draft/unstable directives).

use crate::types::DirectiveFilterMode;

/// Tier classification for a directive.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DirectiveTier {
    /// Core GraphQL spec directives (e.g., @deprecated, @skip, @include)
    Spec,
    /// Federation directives (@key, @shareable, @external, etc.)
    Federation,
    /// Production-ready custom directives (@constraint, @cache, etc.)
    Custom,
    /// Draft / unstable directives being tested
    Draft,
}

/// Classification of a directive by name.
///
/// This is the canonical list used by the filter to determine
/// which directives to include or exclude based on the active mode.
pub fn classify_directive(name: &str) -> DirectiveTier {
    // Strip leading @ if present
    let name = name.trim_start_matches('@');

    match name {
        // GraphQL spec built-ins
        "deprecated" | "skip" | "include" | "specifiedBy" => DirectiveTier::Spec,

        // Federation v2.x directives
        "key" | "shareable" | "external" | "requires" | "provides" | "override"
        | "inaccessible" | "tag" | "interfaceObject" | "authenticated" | "requiresScopes"
        | "policy" | "cost" | "listSize" | "link" | "composeDirective" => DirectiveTier::Federation,

        // Production custom directives
        "constraint" | "cache" | "authorize" | "mask" | "rateLimit" => DirectiveTier::Custom,

        // Unknown / user directives are treated as Custom (include by default)
        _ => DirectiveTier::Custom,
    }
}

/// Determine whether a directive should be included in the output
/// based on the active filter mode.
pub fn should_include_directive(name: &str, mode: &DirectiveFilterMode) -> bool {
    match mode {
        DirectiveFilterMode::All => true,

        DirectiveFilterMode::ViewerFriendly => {
            // Only include spec-level directives in viewer-facing output
            classify_directive(name) == DirectiveTier::Spec
        }

        DirectiveFilterMode::ExcludeDraft => {
            // Include everything except draft directives
            classify_directive(name) != DirectiveTier::Draft
        }

        DirectiveFilterMode::Custom(ref excluded) => {
            let name = name.trim_start_matches('@');
            !excluded
                .iter()
                .any(|d| d.trim_start_matches('@').eq_ignore_ascii_case(name))
        }
    }
}

/// Filter a list of directive strings, keeping only those allowed
/// by the active mode.
pub fn filter_directive_list(directives: &[String], mode: &DirectiveFilterMode) -> Vec<String> {
    directives
        .iter()
        .filter(|d| {
            // Extract directive name from SDL text like "@key(fields: \"id\")"
            let name = d
                .trim_start_matches('@')
                .split(|c: char| c == '(' || c.is_whitespace())
                .next()
                .unwrap_or(d);
            should_include_directive(name, mode)
        })
        .cloned()
        .collect()
}

/// Filter directive tokens from a single line of SDL output.
///
/// Given a line like `  id: ID! @key(fields: "id") @shareable`,
/// strips directives that should be excluded based on the mode.
pub fn filter_line_directives(line: &str, mode: &DirectiveFilterMode) -> String {
    if mode == &DirectiveFilterMode::All {
        return line.to_string();
    }

    let mut result = String::new();
    let mut in_directive = false;
    let mut directive_buffer = String::new();
    let mut paren_depth: i32 = 0;

    for ch in line.chars() {
        if ch == '@' && !in_directive {
            in_directive = true;
            directive_buffer.clear();
            directive_buffer.push(ch);
        } else if in_directive {
            directive_buffer.push(ch);
            if ch == '(' {
                paren_depth += 1;
            } else if ch == ')' {
                paren_depth -= 1;
                if paren_depth == 0 {
                    // End of directive; decide whether to include
                    let name = directive_buffer
                        .trim_start_matches('@')
                        .split(|c: char| c == '(' || c.is_whitespace())
                        .next()
                        .unwrap_or(&directive_buffer);
                    if should_include_directive(name, mode) {
                        result.push_str(&directive_buffer);
                    }
                    in_directive = false;
                    directive_buffer.clear();
                }
            } else if paren_depth == 0 && ch.is_whitespace() {
                // Simple directive without parens ended
                let name = directive_buffer.trim_start_matches('@').trim();
                if should_include_directive(name, mode) {
                    result.push_str(&directive_buffer);
                }
                result.push(ch);
                in_directive = false;
                directive_buffer.clear();
            }
        } else {
            result.push(ch);
        }
    }

    // Handle trailing directive without whitespace
    if in_directive && !directive_buffer.is_empty() {
        let name = directive_buffer
            .trim_start_matches('@')
            .split(|c: char| c == '(' || c.is_whitespace())
            .next()
            .unwrap_or(&directive_buffer);
        if should_include_directive(name, mode) {
            result.push_str(&directive_buffer);
        }
    }

    // Clean up trailing whitespace
    result.trim_end().to_string()
}

/// Apply directive filtering to an entire SDL string.
///
/// Process each line and strip excluded directives.
pub fn filter_sdl_directives(sdl: &str, mode: &DirectiveFilterMode) -> String {
    if mode == &DirectiveFilterMode::All {
        return sdl.to_string();
    }
    sdl.lines()
        .map(|line| filter_line_directives(line, mode))
        .collect::<Vec<_>>()
        .join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_classify_directive() {
        assert_eq!(classify_directive("@deprecated"), DirectiveTier::Spec);
        assert_eq!(classify_directive("@key"), DirectiveTier::Federation);
        assert_eq!(classify_directive("@shareable"), DirectiveTier::Federation);
        assert_eq!(classify_directive("@constraint"), DirectiveTier::Custom);
    }

    #[test]
    fn test_should_include_all_mode() {
        assert!(should_include_directive("@key", &DirectiveFilterMode::All));
        assert!(should_include_directive(
            "@deprecated",
            &DirectiveFilterMode::All
        ));
        assert!(should_include_directive(
            "@custom",
            &DirectiveFilterMode::All
        ));
    }

    #[test]
    fn test_should_include_viewer_friendly() {
        // Only spec directives pass
        assert!(should_include_directive(
            "@deprecated",
            &DirectiveFilterMode::ViewerFriendly
        ));
        assert!(!should_include_directive(
            "@key",
            &DirectiveFilterMode::ViewerFriendly
        ));
        assert!(!should_include_directive(
            "@shareable",
            &DirectiveFilterMode::ViewerFriendly
        ));
    }

    #[test]
    fn test_should_include_exclude_draft() {
        // Everything except draft passes
        assert!(should_include_directive(
            "@key",
            &DirectiveFilterMode::ExcludeDraft
        ));
        assert!(should_include_directive(
            "@deprecated",
            &DirectiveFilterMode::ExcludeDraft
        ));
    }

    #[test]
    fn test_should_include_custom_exclusion() {
        let mode = DirectiveFilterMode::Custom(vec!["@key".to_string()]);
        assert!(!should_include_directive("@key", &mode));
        assert!(should_include_directive("@shareable", &mode));
    }

    #[test]
    fn test_filter_line_directives_viewer_friendly() {
        let line = "  id: ID! @key(fields: \"id\") @deprecated";
        let result = filter_line_directives(line, &DirectiveFilterMode::ViewerFriendly);
        assert!(result.contains("@deprecated"));
        assert!(!result.contains("@key"));
    }

    #[test]
    fn test_filter_line_directives_all_mode_is_noop() {
        let line = "  id: ID! @key(fields: \"id\") @shareable";
        let result = filter_line_directives(line, &DirectiveFilterMode::All);
        assert_eq!(result, line);
    }

    #[test]
    fn test_filter_line_directives_with_arg_directive() {
        let line = "type User @key(fields: \"id\") @shareable {";
        let result = filter_line_directives(line, &DirectiveFilterMode::ViewerFriendly);
        assert!(!result.contains("@key"));
        assert!(!result.contains("@shareable"));
        assert!(result.contains("type User"));
    }

    #[test]
    fn test_filter_sdl_directives_full() {
        let sdl =
            "type User @key(fields: \"id\") {\n  id: ID! @deprecated\n  name: String @shareable\n}";
        let result = filter_sdl_directives(sdl, &DirectiveFilterMode::ViewerFriendly);
        assert!(!result.contains("@key"));
        assert!(!result.contains("@shareable"));
        assert!(result.contains("@deprecated"));
    }

    #[test]
    fn test_filter_directive_list() {
        let directives = vec![
            "@key(fields: \"id\")".to_string(),
            "@deprecated".to_string(),
            "@shareable".to_string(),
        ];
        let result = filter_directive_list(&directives, &DirectiveFilterMode::ViewerFriendly);
        assert_eq!(result.len(), 1);
        assert_eq!(result[0], "@deprecated");
    }
}

File: converters/node/src/directive-filter.ts
/**
 * Directive Filter Framework for GraphQL SDL generation.
 *
 * Configuration-driven filtering of GraphQL directives during SDL output,
 * supporting modes like "viewer-friendly" (strip all infrastructure
 * directives) and "exclude-draft" (omit draft/unstable directives).
 *
 * Mirrors the Rust implementation in `converters/rust/src/directive_filter.rs`.
 */

export type DirectiveTier = "spec" | "federation" | "custom" | "draft";

/**
 * Public filter mode. Uses the same enum values as the GraphQL API
 * schema (ALL, VIEWER_FRIENDLY, EXCLUDE_DRAFT) plus a `Custom` variant
 * with a list of directive names to exclude.
 */
export type DirectiveFilterMode =
  "ALL" | "VIEWER_FRIENDLY" | "EXCLUDE_DRAFT" | { CUSTOM: string[] };

/** Classify a directive by its tier. */
export function classifyDirective(name: string): DirectiveTier {
  const cleanName = name.startsWith("@") ? name.slice(1) : name;

  // GraphQL spec built-ins
  if (
    cleanName === "deprecated" ||
    cleanName === "skip" ||
    cleanName === "include" ||
    cleanName === "specifiedBy"
  ) {
    return "spec";
  }

  // Federation v2.x directives
  const federationDirectives = new Set([
    "key",
    "shareable",
    "external",
    "requires",
    "provides",
    "override",
    "inaccessible",
    "tag",
    "interfaceObject",
    "authenticated",
    "requiresScopes",
    "policy",
    "cost",
    "listSize",
    "composeDirective",
  ]);
  if (federationDirectives.has(cleanName)) {
    return "federation";
  }

  // Production custom directives
  const customDirectives = new Set([
    "constraint",
    "cache",
    "authorize",
    "mask",
    "rateLimit",
  ]);
  if (customDirectives.has(cleanName)) {
    return "custom";
  }

  // Unknown → treat as custom
  return "custom";
}

/** Determine if a directive should be included given the active mode. */
export function shouldIncludeDirective(
  name: string,
  mode: DirectiveFilterMode,
): boolean {
  if (mode === "ALL") {
    return true;
  }

  if (mode === "VIEWER_FRIENDLY") {
    return classifyDirective(name) === "spec";
  }

  if (mode === "EXCLUDE_DRAFT") {
    return classifyDirective(name) !== "draft";
  }

  // Custom exclusion list
  if (typeof mode === "object" && "CUSTOM" in mode) {
    const cleanName = (
      name.startsWith("@") ? name.slice(1) : name
    ).toLowerCase();
    return !mode.CUSTOM.some((d) => {
      const dName = (d.startsWith("@") ? d.slice(1) : d).toLowerCase();
      return dName === cleanName;
    });
  }

  return true;
}

/** Filter an array of directive strings based on the mode. */
export function filterDirectiveList(
  directives: string[],
  mode: DirectiveFilterMode,
): string[] {
  return directives.filter((d) => {
    const name = d.trimStart().replace(/^@/, "").split(/[(\s]/).shift() ?? d;
    return shouldIncludeDirective(name, mode);
  });
}

/** Filter directive tokens from a single line of SDL output. */
export function filterLineDirectives(
  line: string,
  mode: DirectiveFilterMode,
): string {
  if (mode === "ALL") {
    return line;
  }

  let result = "";
  let inDirective = false;
  let directiveBuffer = "";
  let parenDepth = 0;

  for (const ch of line) {
    if (ch === "@" && !inDirective) {
      inDirective = true;
      directiveBuffer = "";
      directiveBuffer += ch;
    } else if (inDirective) {
      directiveBuffer += ch;
      if (ch === "(") {
        parenDepth++;
      } else if (ch === ")") {
        parenDepth--;
        if (parenDepth === 0) {
          const name =
            directiveBuffer.replace(/^@/, "").split(/[(\s]/).shift() ??
            directiveBuffer;
          if (shouldIncludeDirective(name, mode)) {
            result += directiveBuffer;
          }
          inDirective = false;
          directiveBuffer = "";
        }
      } else if (parenDepth === 0 && /\s/.test(ch)) {
        const name = directiveBuffer.replace(/^@/, "").trim();
        if (shouldIncludeDirective(name, mode)) {
          result += directiveBuffer;
        }
        result += ch;
        inDirective = false;
        directiveBuffer = "";
      }
    } else {
      result += ch;
    }
  }

  if (inDirective && directiveBuffer) {
    const name =
      directiveBuffer.replace(/^@/, "").split(/[(\s]/).shift() ??
      directiveBuffer;
    if (shouldIncludeDirective(name, mode)) {
      result += directiveBuffer;
    }
  }

  return result.replace(/\s+$/, "");
}

/** Apply directive filtering to an entire SDL string. */
export function filterSdlDirectives(
  sdl: string,
  mode: DirectiveFilterMode,
): string {
  if (mode === "ALL") {
    return sdl;
  }
  return sdl
    .split("\n")
    .map((line) => filterLineDirectives(line, mode))
    .join("\n");
}

export const FEDERATION_VERSION = "2.9";

export const FEDERATION_IMPORTS = [
  "@key",
  "@shareable",
  "@external",
  "@provides",
  "@requires",
  "@override",
  "@inaccessible",
  "@tag",
  "@interfaceObject",
  "@authenticated",
  "@requiresScopes",
  "@policy",
  "@cost",
  "@listSize",
];

/** Build the @link extend schema directive line. */
export function federationLinkDirective(): string {
  const imports = FEDERATION_IMPORTS.map((i) => `"${i}"`).join(", ");
  return `extend schema @link(url: "https://specs.apollo.dev/federation/v${FEDERATION_VERSION}", import: [${imports}])`;
}

/** Complete Federation v2.9 directive SDL definitions. */
export const FEDERATION_DIRECTIVES_SDL = `
scalar FieldSet
scalar link__Import

enum link__Purpose {
  """
  \`SECURITY\` features provide metadata necessary to securely resolve fields.
  """
  SECURITY

  """
  \`EXECUTION\` features provide metadata necessary for operation execution.
  """
  EXECUTION
}

directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
directive @requires(fields: FieldSet!) on FIELD_DEFINITION
directive @provides(fields: FieldSet!) on FIELD_DEFINITION
directive @external(reason: String) on FIELD_DEFINITION | OBJECT
directive @shareable on FIELD_DEFINITION | OBJECT
directive @link(url: String!, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
directive @tag(name: String!) repeatable on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
directive @override(from: String!, label: String) on FIELD_DEFINITION
directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
directive @interfaceObject on OBJECT
directive @composeDirective(name: String) repeatable on SCHEMA
directive @authenticated on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM
directive @requiresScopes(scopes: [[String!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM
directive @policy(policies: [[String!]!]!) on FIELD_DEFINITION | OBJECT | INTERFACE | SCALAR | ENUM
directive @cost(weight: Int!) on ARGUMENT_DEFINITION | ENUM | FIELD_DEFINITION | INPUT_FIELD_DEFINITION | OBJECT | SCALAR
directive @listSize(assumedSize: Int, slicingArguments: [String!], sizedFields: [String!], requireOneSlicingArgument: Boolean = true) on FIELD_DEFINITION

# Custom validation and performance directives
directive @constraint(pattern: String, min: Int, max: Int, minLength: Int, maxLength: Int) on FIELD_DEFINITION | ARGUMENT_DEFINITION
directive @cache(ttl: Int, scope: String) on FIELD_DEFINITION | OBJECT
directive @authorize(requires: String, roles: [String], scopes: [String]) on FIELD_DEFINITION | OBJECT | INTERFACE
directive @mask(pattern: String, if: String, value: String, maskLevel: String, maskFor: [String]) on FIELD_DEFINITION
directive @rateLimit(limit: Int, window: String, roles: [String]) on FIELD_DEFINITION
`;

/** Check whether an SDL string already contains federation directive definitions. */
export function sdlHasFederationDirectives(sdl: string): boolean {
  return (
    sdl.includes("directive @key") ||
    sdl.includes("directive @shareable") ||
    sdl.includes("directive @link")
  );
}

/** Inject federation directive definitions into an SDL string if missing. */
export function ensureFederationDirectives(sdl: string): string {
  if (sdlHasFederationDirectives(sdl)) {
    return sdl;
  }
  // Remove any existing @link line to avoid conflicts
  const cleaned = sdl
    .split("\n")
    .filter((l) => !l.trim().startsWith("extend schema @link"))
    .join("\n");
  return `${FEDERATION_DIRECTIVES_SDL.trim()}\n${cleaned.trim()}`;
}
